errbot/backends/irc.py

```diff
@@ -416,6 +430,21 @@ def on_dcc_connect(self, dcc, event):
     def on_dcc_disconnect(self, dcc, event):
         self.transfers.pop(dcc)
 
+    def on_endofnames(self, c, e):
+        # The e.arguments[0] contains the channel name.
+        # We filter that to avoid a misfire of the event.
+        room_name = e.arguments[0]
+        if room_name in self.__recently_joined_to:
```

Comment:
Specifically, there's a chance that `on_endofnames` is called right around the same time multiple times from different threads (because of lot of these callbacks are handled in their own thread). What might happen is that they both attempt to update the set at the same time, so you have to guard against this.

There's `threading.Lock()` for this. I'd be happy to explain how to use it if you have never used it before.